update-engine/src/buffer.rs

```diff
@@ -0,0 +1,1011 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+use std::collections::{hash_map, HashMap, VecDeque};
+
+use derive_where::derive_where;
+use either::Either;
+use petgraph::prelude::*;
+
+use crate::{
+    events::{
+        Event, EventReport, ProgressEvent, ProgressEventKind, StepEvent,
+        StepEventKind, StepEventPriority,
+    },
+    ExecutionId, StepSpec,
+};
+
+/// A receiver for events that provides a pull-based model with periodic
+/// reports.
+///
+/// By default, the update engine provides a *push-based model* where you are
+/// notified of new events as soon as they come in. The event buffer converts
+/// events into a *pull-based model*, where periodic, serializable
+/// [`EventReport`]s can be generated.
+///
+/// # Features
+///
+/// The buffer is responsible for tracking step and progress events as they
+/// come in. The buffer can:
+///
+/// * Receive events and update its internal state based on them.
+/// * Discard progress events that are no longer useful.
+/// * Cap the number of [low-priority events](StepEventPriority::Low) such that
+///   older events are dropped.
+///
+/// The buffer is currently resilient against:
+///
+/// * duplicated and dropped progress events
+/// * duplicated step events
+/// * dropped *low-priority* step events
+///
+/// The buffer is currently *not* resilient against:
+/// * dropped *high-priority* step events
+/// * reordered progress or step events
+///
+/// These cases can be handled on a best-effort basis in the future at some cost
+/// to complexity, if required.
+#[derive_where(Clone, Debug)]
+pub struct EventBuffer<S: StepSpec> {
```

Comment:
BTW I didn't really have a better name for this. I originally called it `EventCoalescer` but I felt it was too wordy. The main goal of this is to track the state of an update engine running elsewhere.

Reply:
I like `EventBuffer` more than `EventCoalescer` too. Maaaaybe something like `EventCombiner` or `EventReportCombiner`? I'm fine with `EventBuffer` though.